python/extensions/pip.bzl

```diff
@@ -14,63 +14,140 @@
+            isolated = use_isolated(module_ctx, pip_attr),
+            extra_pip_args = extra_pip_args,
+            download_only = pip_attr.download_only,
+            pip_data_exclude = pip_attr.pip_data_exclude,
+            enable_implicit_namespace_pkgs = pip_attr.enable_implicit_namespace_pkgs,
+            environment = pip_attr.environment,
+        )
+
+        if whl_name not in whl_map[hub_name]:
+            whl_map[hub_name][whl_name] = {}
+
+        whl_map[hub_name][whl_name][pip_attr.python_version] = pip_name + "_"
+
+    return whl_map
+
 def _pip_impl(module_ctx):
+    root_hubs = {}
+    submodules_hubs = {}
+    hub_whl_map = {}
+
+    # TODO we need to check for duplicate python versions in the same hub
+    # TODO we need to check for duplicate python hub names in submodules
+    # TODO improve failure messages
     for mod in module_ctx.modules:
-        for attr in mod.tags.parse:
-            requrements_lock = locked_requirements_label(module_ctx, attr)
-
-            # Parse the requirements file directly in starlark to get the information
-            # needed for the whl_libary declarations below. This is needed to contain
-            # the pip_repository logic to a single module extension.
-            requirements_lock_content = module_ctx.read(requrements_lock)
-            parse_result = parse_requirements(requirements_lock_content)
-            requirements = parse_result.requirements
-            extra_pip_args = attr.extra_pip_args + parse_result.options
-
-            # Create the repository where users load the `requirement` macro. Under bzlmod
-            # this does not create the install_deps() macro.
-            pip_repository_bzlmod(
-                name = attr.name,
-                repo_name = attr.name,
-                requirements_lock = attr.requirements_lock,
-                incompatible_generate_aliases = attr.incompatible_generate_aliases,
+        for pip_attr in mod.tags.parse:
+            if mod.is_root:
+                if pip_attr.hub_name in root_hubs:
+                    root_hubs[pip_attr.hub_name].append(pip_attr)
+                else:
+                    root_hubs[pip_attr.hub_name] = [pip_attr]
+            elif pip_attr.hub_name in submodules_hubs:
+                submodules_hubs[pip_attr.hub_name].append(pip_attr)
+            else:
+                submodules_hubs[pip_attr.hub_name] = [pip_attr]
+
+    for hub_name, pip_attr in root_hubs.items():
+        for pip in pip_attr:
+            # TODO we need to check to see if we have the same python version
+            # in a different pip call.
+            hub_whl_map = _create_pip(module_ctx, pip, hub_name, hub_whl_map)
+
+    for hub_name, pip_attr in submodules_hubs.items():
+        if hub_name in root_hubs:
+            # TODO clean up the warn statement like we are doing in toolchains
+            _print_warn("Not creating pip with the hub_name {}, same hub name found in the root module.".format(hub_name))
```

Comment:
So then the user will have to modify sub modules

Reply:
Unfortunately, yes, but there isn't much we can do about it. The only options are:
1. The user manually picks a globally unique hub name
2. We automatically put the module name in the hub name to force it to be globally unique, and users have the option of using repo aliases.
3. The user writes their own extension that does what pip.parse does (so that the repos end up in their module's namespace, not rules_python), probably based on helpers we provided (this could be as simple as a user extension just loading our extension and re-exporting the name, since the use_extension() args define identity).
4. We wait for the `is_isolated` feature that allows a particular extension usage to have its own namespace.

(1) is today. I'm mixed about (2) and (3). (4) is ideal, but doesn't exist yet.